ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,369 @@
-
-        if len(file) > 0:
-            self.fileContent = self._readFileContent(file)
-            self.instructionList, self.referenceDict, self.finalSize = self._parseCodeFile()
-        return
+    relativeAddressCounter = 0                # Used to determine memory address of a label at a given point in the file
+    lineno = 0                                # Used to pinpoint where ValueErrors are raised
 
-    def _readFileContent(self, file: str=""):
+    def parse(self, text):
         """
-        This will simply read the content into the local buffer for it to be parsed
-        :param file: String, the name of the file that needs to be read in to build the code
-        :return: list: Content of the file, line by line.
+        This is the main driver method for the parser. It takes in an individual line of code, parses the info and
+        sends the fully built binary code to the assembler, or returns a memory reference with the appropriate offset.
+        :param text: Str, The line of text as read from the file
+        :return: The encoded binary instruction for the line, the current offset, and the label flag used by assembler
         """
 
-        content = []
-        codeFile = open(file, "r")
-        origContent = codeFile.readlines()
-        codeFile.close()
+        line = text.split()                     # We split the line with space as a delimiter. Each token is an argument
+        instruction = b""                       # Instruction will be written as a bytestring
+        arglist = []                            # We keep each argument in a list to evaluate later
+        found = False                           # Flag to indicate if instruction was found in state list
+        form = "Ins"                            # We assume the first argument is an instruction
+        labelFlag = 0                           # Used by the assembler to determine what we're returning
+        dataIdentifier = line[0].upper()        # used for special identifiers such as .dataAlpha, comments etc.
+        self.lineno += 1
+
+        # First off, we need to determine if this line has a label or global label to be referenced
+        # If so, we can simply return to the Assembler class with the label, the offset, and the appropriate flag
+        identifierMatch = re.search(r'(\.(\w*))|(\w*:)|(;)', dataIdentifier)
+
+        if identifierMatch:
+            # We had a match for identifier patterns, so we assume there's no instruction.
+            instruction, labelFlag = self._evaluateIndicatorData(text, dataIdentifier, line, instruction, labelFlag)
+            return instruction, self.relativeAddressCounter, labelFlag
+
+        elif line[0].upper() in INSTRUCTION_LIST:
+            # Next we check if the first item on the line is an instruction
+            # If we make it to this line, there were no data indicators (.dataAlpha, comments, labels etc.)
+            line = text.upper()
```

Comment:
Why not do this in before the first call to split() and do a single split() ?

Reply:
turns out, it already is done before the first call! Just an extra line that didn't need to be there.

Reply:
Upon review, this was necessary because converting to upper isn't possible with a split line. I need to preserve the text's line as is before converting to upper for .dataAlpha fields.